community/kernel/src/main/java/org/neo4j/helpers/SortedLongArrayUtil.java

```diff
@@ -0,0 +1,251 @@
+            else
+            {
+                r++;
+            }
+        }
+
+        assert cursor == intersect.length;
+        return intersect;
+    }
+
+    /**
+     * Compute the symmetric difference (set XOR basically) of two sorted long array sets.
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return the union, which is NOT sorted
+     */
+    public static long[] symmetricDifference( long[] left, long[] right )
+    {
+        if ( left == null || right == null )
+        {
+            return left == null ? right : left;
+        }
+
+        long uniqueCounts = countUnique( left, right );
+        if ( uniqueCounts == 0 ) // complete intersection
+        {
+            return EMPTY;
+        }
+
+        long[] difference = new long[left( uniqueCounts ) + right( uniqueCounts )];
+
+        int cursor = 0;
+        int l = 0;
+        int r = 0;
+        while ( l < left.length && r < right.length )
+        {
+            if ( left[l] == right[r] )
+            {
+                l++;
+                r++;
+            }
+            else if ( left[l] < right[r] )
+            {
+                difference[cursor++] = left[l];
+                l++;
+            }
+            else
+            {
+                difference[cursor++] = right[r];
+                r++;
+            }
+        }
+        while ( l < left.length )
+        {
+            difference[cursor++] = left[l];
+            l++;
+        }
+        while ( r < right.length )
+        {
+            difference[cursor++] = right[r];
+            r++;
+        }
+
+        assert cursor == difference.length;
+        return difference;
+    }
+
+    /**
+     * Compute the number of unique values in two sorted long array sets
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return int pair packed into long
+     */
+    public static long countUnique( long[] left, long[] right )
+    {
+        int l = 0;
+        int r = 0;
+        int uniqueInLeft = 0;
+        int uniqueInRight = 0;
+        while ( l < left.length && r < right.length )
+        {
+            if ( left[l] == right[r] )
+            {
+                l++;
+                r++;
+            }
+            else if ( left[l] < right[r] )
+            {
+                uniqueInLeft++;
+                l++;
+            }
+            else
+            {
+                uniqueInRight++;
+                r++;
+            }
+        }
+        uniqueInLeft += left.length - l;
+        uniqueInRight += right.length - r;
+        return intPair( uniqueInLeft, uniqueInRight );
+    }
+
+    public static long intPair( int left, int right )
+    {
+        return (long)left << (Integer.BYTES * 8) | right;
+    }
+
+    public static int left( long pair )
+    {
+        return (int)(pair >> (Integer.BYTES * 8));
+    }
+
+    public static int right( long pair )
+    {
+        return (int)pair;
```

Comment:
This method is wrong since the conversion from `long` to int is sign-preserving.

A correct implementation would be `return (int)(pair & 0xFFFF_FFFFL);`.